spec/system/consumer/split_checkout_spec.rb

```diff
@@ -708,6 +708,93 @@
         end
       end
 
+      describe "vouchers" do
+        context "with no voucher available" do
+          before do
+            visit checkout_step_path(:payment)
+          end
+
+          it "doesn't show voucher input" do
+            expect(page).not_to have_content "Apply voucher"
+          end
+        end
+
+        context "with voucher available" do
+          let!(:voucher) { Voucher.create(code: 'some_code', enterprise: distributor) }
+
+          before do
+            visit checkout_step_path(:payment)
+          end
+
+          it "shows voucher input" do
+            expect(page).to have_content "Apply voucher"
+          end
+
+          describe "adding voucher to the order" do
+            shared_examples "adding voucher to the order" do
+              before do
+                fill_in "Enter voucher code", with: voucher.code
+                click_button("Apply")
+              end
+
+              it "adds a voucher to the order" do
+                expect(page).to have_content("$10.00 Voucher")
+                expect(order.reload.voucher_adjustments.length).to eq(1)
+              end
+            end
+
+            it_behaves_like "adding voucher to the order"
+
+            context "when voucher covers more then the order total" do
+              before do
+                order.total = 6
+                order.save!
+              end
+
+              it_behaves_like "adding voucher to the order"
+
+              it "shows a warning message" do
+                fill_in "Enter voucher code", with: voucher.code
+                click_button("Apply")
+
+                expect(page).to have_content(
+                  "Your voucher value is more than your order. " \
+                  "By using this voucher you are forfeiting the remaining value."
+                )
+              end
+            end
+
+            context "voucher doesn't exist" do
+              it "show an error" do
+                fill_in "Enter voucher code", with: "non_code"
+                click_button("Apply")
+
+                expect(page).to have_content("Voucher Not found")
+              end
+            end
+          end
+
+          describe "removing voucher from order" do
+            before do
+              voucher.create_adjustment(voucher.code, order)
```

Comment:
Is there a reason we pass the code on? Doesn't the voucher know it?

Reply:
For consistency, I kept the same signature as here https://github.com/openfoodfoundation/openfoodnetwork/blob/d2fbb0bde2db2d230fb79dbba302f2d239060fe9/app/models/concerns/calculated_adjustments.rb#L43

So here `voucher.code` is the label that is going to be applied to the adjustment created, but you are right that the voucher does know it's own code :)

Reply:
Ah, now I remember that I came to the same conclusion in a previous review. It's the label which *could* be different. All good.